### ISSUE TITLE
Gradient Clipping 

### ISSUE THREAD

**Issue by neural-everything** [lululxvi/deepxde#1197]
Hi Lu, I am working on a set of PDEs that are extremely unstable. I am experiencing quite a lot of gradient explosion. Is there a way to implement clipping with deepxde ? Thanks in advance

**Comment by lululxvi** [comment]
Then you need to modify the source code.

**Comment by kyouma** [comment]
As for Pytorch backend, you can register a backward hook for your **network** (for example, deepxde.nn.FNN) with gradient clipping for each layer separately: https://stackoverflow.com/a/54816498/10901462
However, this approach is not equal to clipping the whole gradient norm and, according to the asnwer's comments, even to clipping gradients for each layer separately because of different execution order.

**Comment by neural-everything** [comment]
Hi again. @kyouma, @lululxvi thank you for your replies. I managed to do it by adding: 
`flat_grads = tf.concat([tf.reshape(grd,[-1,]) for grd in grads],axis=0)`
`if tf.math.greater(tf.norm(flat_grads),norm):`
  `                  flat_grads = tf.clip_by_norm(flat_grads,norm)`
`grads = tf_unflatten(flat_grads,trainable_variables) `
before
`opt.apply_gradients(zip(grads, trainable_variables))` in the `train_step()` function within `compile_tensorflow()`
